### ISSUE TITLE
Support Apache TinkerPop and the Gremlin traversal machine and language.

### ISSUE THREAD

**Issue by okram** [kakao/s2graph#175]
Are there plans to support Apache TinkerPop (http://tinkerpop.apache.org)? If you do, you get Gremlin OLTP and OLAP for free. For your specific optimizations (fast breadth-first search), you could write a `ProivderOptimizationTraversalStrategy` to compile Gremlin to leverage S2Graph's unique capabilities. This is how Titan is able to leverage "vertex-centric indices" by having their own strategy for rewriting particular steps to use their local index system. There are alot of other goodies you get for free like GremlinServer (REST/binary server with metrics/etc.), support for languages not on the JVM, Gephi visualization, Hadoop/Spark/Giraph support.

If you need some help/direction, we are more than happy to advise.

Reading material:
- http://www.datastax.com/dev/blog/the-benefits-of-the-gremlin-graph-traversal-machine
- http://arxiv.org/abs/1508.03843
- http://www.slideshare.net/slidarko/acm-dbpl-keynote-the-graph-traversal-machine-and-language (see slides 128-140 and see how you get Spark/Giraph/etc. OLAP support for free).
- http://tinkerpop.incubator.apache.org/docs/3.1.0-incubating/#gremlin-server
- http://tinkerpop.incubator.apache.org/docs/3.1.0-incubating/#traversalstrategy (explaining how strategies allow graph system providers to alter Gremlin compilation to use their system's unique features).

**Comment by SteamShon** [comment]
@okram 
absolutely! we know that TinkerPop provide lots of nice goodies. we have plan to support TinkerPop3 soon and I think I will need help for that when we implement ProivderOptimizationTraversalStrategy like you mention. 

We are currently refactoring code bases so we can easily support TinkerPop. the old code base is tightly coupled with HBase specific codes. Once we finished refactoring(I could say in 2~3 weeks), we will definitely work on implementing ThinkerPop.

**Comment by wikier** [comment]
:+1:

**Comment by spmallette** [comment]
:+1:

**Comment by pluradj** [comment]
+1

**Comment by okram** [comment]
When you release a version of S2Graph that has TinkerPop, be sure to ping us (https://twitter.com/apachetinkerpop) and we will add S2Graph to our homepage (http://tinkerpop.apache.org/) and promote it on Twitter/etc.

Also, if you are feeling really adventurous, it would be neat to have a blog post talking about S2Graph + TinkerPop. We can then promote it around as well to help get more traffic to both our respective projects.

**Comment by wishoping** [comment]
:+1:
